### PR TITLE
Rest controller

### PR DESCRIPTION
**Pull request by gaurav-dogra**
Wrote Test for RestController in WebLayerTest.java, using mocking the server does not have to start while testing. most of other changes is refactoring.

### CODE REVIEW

File: src/main/java/gmailgdogra/OutputRow.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package gmailgdogra;
+
+public class OutputRow {
+    private final String firstName;
+    private final String lastName;
+    private final String swipeTime;
+    private final String deviceName;
+
+    public static OutputRow of(String firstName, String lastName, String swipeTime, String deviceName) {
```
Comment (alfabravo2013) [inline src/main/java/gmailgdogra/OutputRow.java:9]: If you use this factory method, it'd be a good idea to add assertions that the args are not null

File: src/main/java/gmailgdogra/ReadXlsxService.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+
+@Service
+public class ReadXlsxService {
+
+    public static String EXCEL_FILE_TYPE = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";
+
+    public static List<SwipeRecord> parse(InputStream inputStream) throws IOException {
```
Comment (alfabravo2013) [inline src/main/java/gmailgdogra/ReadXlsxService.java:15]: Where do you actually catch this IOException? Also, ask around what is the best practic: return an empty list or propagate the exception.

File: src/main/java/gmailgdogra/SwipeProcessor.java [hunk 3]
```diff
@@ -39,56 +34,51 @@
...
-                .filter(swipeRecord -> {
+    public static SwipeRecord getFirstSwipeIn(Officer officer, List<SwipeRecord> allSwipes) {
+        Location loc = officer.getLocation();
+        return allSwipes.stream()
+                .filter(swipe -> officer.getFullName().equals(swipe.getFullName()))
+                .filter(swipe -> allSwipeInDevices.get(loc).contains(swipe.getDeviceName()))
+                .filter(swipe -> {
                     if (officer.isDayShift()) {
-                        return swipeRecord.getSwipeTime().getHour() <= 12; // touch-in is before 12 noon
+                        return swipe.getSwipeTime().getHour() <= 12; // touch-in is before 12 noon
```
Comment (alfabravo2013) [inline src/main/java/gmailgdogra/SwipeProcessor.java:57]: I'm not sure about your business logic. In your case, 12:30 will be treated as "before 12 noon". Is it correct?

File: src/test/java/gmailgdogra/ExtractOfficersTest.java [hunk 4]
```diff
@@ -24,7 +29,18 @@
...
     void from() throws IOException {
-        List<SwipeRecord> swipeRecords = ReadXlsx.parse("src/main/resources/testFile.xlsx");
+        Path path = Paths.get("src/main/resources/testFile.xlsx");
+        String name = "testFile.xlsx";
+        String originalFileName = "testFile.xlsx";
+        String contentType = "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet";
+        byte[] content = null;
+        try {
+            content = Files.readAllBytes(path);
+        } catch (final IOException ignored) {
```
Comment (alfabravo2013) [inline src/test/java/gmailgdogra/ExtractOfficersTest.java:39]: Ignored exceptions bother people. You coul've failed the test with a meaningful message instead.

**Review by alfabravo2013 — APPROVED** [review]
Good job. I see you added CI on GitHub, but it fails to build your app. If you find any good resource on how to setup GitHub Actions, tell me.

File: src/main/java/gmailgdogra/OutputRow.java (hunk 1 above)
Comment (gaurav-dogra) [reply src/main/java/gmailgdogra/OutputRow.java:9]: agree

File: src/main/java/gmailgdogra/ReadXlsxService.java (hunk 2 above)
Comment (gaurav-dogra) [reply src/main/java/gmailgdogra/ReadXlsxService.java:15]: I was catching this in RestController
```
catch (Exception e) {
                return ResponseEntity.status(HttpStatus.EXPECTATION_FAILED)
                        .body(new ResponseMessage("Unable to parse file: " + file.getName()));
            }
```
And there is a test for this exception being thrown

File: src/main/java/gmailgdogra/SwipeProcessor.java (hunk 3 above)
Comment (gaurav-dogra) [reply src/main/java/gmailgdogra/SwipeProcessor.java:57]: Let me explain

I am trying to find if it is a day shift(0600 - 1800) or a night shift(1800-0600). Almost every time touch-In happen between 0500-0700 for morning shift, and 1700-1900 for night shift. for clarity I am going to change the timing to most likely scenario, i.e. 0500-0800, 1700-2000

Comment (gaurav-dogra) [reply src/main/java/gmailgdogra/SwipeProcessor.java:57]: updated the comments for more clarity


```
  if (officer.isDayShift()) {
                        return swipe.getSwipeTime().getHour() <= 12; // touch-in is before 12:59 afternoon
                    } else {
                        return swipe.getSwipeTime().getHour() > 12; // touch-in on or after 13:00
                    }

```

File: src/test/java/gmailgdogra/ExtractOfficersTest.java (hunk 4 above)
Comment (gaurav-dogra) [reply src/test/java/gmailgdogra/ExtractOfficersTest.java:39]: changed to 
```
try {
            content = Files.readAllBytes(path);
        } catch (IOException e) {
            Assertions.fail("Unable to read file at " + path);
        }
```

**Pull request merged**